File: go/iterator_for_combination.go
package leetcode

type CombinationIterator struct {
	idx          int
	combinations []string
}

func Constructor(characters string, combinationLength int) CombinationIterator {
	iterator := CombinationIterator{
		idx:          0,
		combinations: make([]string, 0),
	}
	var backtrack func(curr string, chars []rune)

	backtrack = func(curr string, chars []rune) {
		if len(curr) == combinationLength {
			iterator.combinations = append(iterator.combinations, curr)
			return
		}

		for i, c := range chars {
			backtrack(curr+string(c), chars[i+1:])
		}
	}
	backtrack("", []rune(characters))
	return iterator
}

func (this *CombinationIterator) Next() string {
	this.idx += 1
	return this.combinations[this.idx-1]
}

func (this *CombinationIterator) HasNext() bool {
	return this.idx < len(this.combinations)
}

File: python/iterator-for-combination.py
class CombinationIterator:

    def __init__(self, characters: str, combinationLength: int):
        self.chars = list(characters)
        self.length = combinationLength
        self.idx = 0
        self.combinations = []
        
        def backtrack(idx, partial_res):
            if len(partial_res) == self.length:
                self.combinations.append("".join(partial_res))
                return
            
            for i in range(idx, len(self.chars)):
                partial_res.append(self.chars[i])
                backtrack(i+1, partial_res)
                partial_res.pop()
        backtrack(0, [])
        
    def next(self) -> str:
        self.idx +=1
        return self.combinations[self.idx-1]

    def hasNext(self) -> bool:
        return self.idx < len(self.combinations)